test/autograd/test_fallback.py

```diff
@@ -315,6 +315,45 @@ def test_post_autograd_returns_leaf(self, mode):
             with self._check_ctx(mode):
                 z.sum().backward()
 
+    @parametrize("mode", ("nothing", "warn"))
+    def test_undefined_inputs_outputs(self, mode):
+        with autograd_fallback_mode(mode):
+            lib = self.get_lib()
+            lib.define("foo(Tensor a, Tensor b) -> (Tensor, Tensor)")
+            op = self.get_op("foo")
+
+            def foo_impl(a, b):
+                return None, b.clone()
+
+            lib.impl("foo", foo_impl, "CPU")
+
+            x = torch.randn(3, requires_grad=True)
+            # NB: PyTorch dispatcher treats "None" as undefined Tensor.
+            y, z = op(None, x)
+            with self._check_ctx(mode):
+                z.sum().backward()
+
+    @parametrize("mode", ("nothing", "warn"))
+    def test_undefined_grads(self, mode):
+        with autograd_fallback_mode(mode):
+            lib = self.get_lib()
+            lib.define("foo(Tensor a, Tensor b) -> (Tensor, Tensor)")
+            op = self.get_op("foo")
+
+            def foo_impl(a, b):
+                return (
+                    torch._C._functions.UndefinedGrad()(a),
+                    torch._C._functions.UndefinedGrad()(b),
+                )
```

Comment:
Are you trying to test the case where your outputs requires_grad=True?  In that case your hooks are registered on the outputs of this function, so they should fire before this happens?

Reply:
Good point